client/peerset/src/lib.rs

```diff
@@ -190,11 +232,10 @@ pub struct PeersetConfig {
 pub struct Peerset {
 	/// Underlying data structure for the nodes's states.
 	data: peersstate::PeersState,
-	/// If true, we only accept reserved nodes.
-	reserved_only: bool,
-	/// Lists of nodes that don't occupy slots and that we should try to always be connected to.
-	/// Is kept in sync with the list of reserved nodes in [`Peerset::data`].
-	priority_groups: HashMap<String, HashSet<PeerId>>,
+	/// For each set, lists of nodes that don't occupy slots and that we should try to always be
+	/// connected to, and whether only reserved nodes are accepted. Is kept in sync with the list
+	/// of non-slot-occupying nodes in [`Peerset::data`].
+	reserved_nodes: Vec<(HashSet<PeerId>, bool)>,
```

Comment:
Instead of treating the `SetId` as an index into the `Vec`, why not use a `HashMap` instead? I would expect this to be less error prone, especially since anyone could create their own via `SetId::from(_)`.
```suggestion
	reserved_nodes: HashMap<SetId, (HashSet<PeerId>, bool)>,
```

Reply:
I don't really understand how that is better.
The valid values for `SetId` range from `0` to `config.sets.len()`. If we put a `HashMap` instead, well, the valid values would remain the same.

Note that managing `SetId`s is entirely scoped to `protocol.rs`. Nothing outside of `protocol.rs` is supposed to generate a `SetId`, apart from `behaviour.rs` that converts them back and forth with a `usize`.

Reply:
Both approaches seem a bit brittle off the top of my head, e.g. for the `Vec` approach with `assert!(set < self.sets.len());`, but also for the `HashMap` approach with the `unwrap` calls needed if an entry doesn't exist.

Whether `HashMap` is better than a `Vec` is quite subjective. I am fine with this given that `SetId` is not exposed outside of `client/network` and given the recent changes through https://github.com/paritytech/substrate/pull/7700/commits/ebaeaab210fdea2f66ae217ee7b8d2f57e7ea38c moving the first two set ids to constants.